### ISSUE TITLE
Maybe something changed recently to cause CategoricalLogit to no longer validate -inf in argument?

### ISSUE THREAD

**Issue by jaberg** [pyro-ppl/numpyro#2282]
### Bug Description

CategoricalLogits fails (maybe after recently-changed code?) when one of the logits is -inf.

### Steps to Reproduce
`CategoricalLogits(logits=[-float('inf'), 0.0]).sample()`

### Expected Behavior
The expression above should return a sample of [1] with probability 1.

There's a comment in the code near the Real constraint that concedes it might not be right to disallow +- inf. I think in this case the CategoricalLogits constraint should somehow admit -inf but not +inf. There are a few ways to do it, and I could maybe implement one as a PR.  It might be slow as I'd be first-time contributor, but I could give it a go. What would be the best fix?

**Comment by juanitorduz** [comment]
Thanks for the report, and yes, you found a real regression.

## Why it fails

The math is fine. Only the argument check rejects it. With validation disabled, `CategoricalLogits([-inf, 0.])` samples `1` with probability 1, gives `log_prob == [-inf, 0.]`, `probs == [0., 1.]`, and finite gradients.

What rejects it is the constraint you spotted:

```python
# numpyro/distributions/constraints.py
class _Real(_SingletonConstraint):
    def __call__(self, x):
        # XXX: consider to relax this condition to [-inf, inf] interval
        return (x == x) & (x != float("inf")) & (x != float("-inf"))
```

`CategoricalLogits.arg_constraints` is `{"logits": constraints.real_vector}`, so one `-inf` entry invalidates the whole vector.

That constraint has always been this strict. What changed recently is that argument validation started running **by default**: #2095 flipped `_VALIDATION_ENABLED` from `False` to `True`, released in 0.20.0. Before that the check existed but never ran unless you called `numpyro.enable_validation()`.

Two details that show the check is wrong rather than protective:

1. The `probs` parameterization already accepts this distribution, and even hands out the very vector the `logits` parameterization refuses:

   ```python
   >>> dist.CategoricalProbs(jnp.array([0.0, 1.0])).logits
   Array([-inf,   0.], dtype=float32)
   ```

2. It is eager-only. Inside `jit` the check sees a tracer and is skipped, so the same code raises in a traced model and passes under `jax.jit`.

The same regression hits other parameters, notably `TruncatedNormal(0., 1., low=-jnp.inf)`, which also raises since 0.20.0 even though it is just an untruncated normal.

## Suggested fix

Please do go for it, and here is the shape I would suggest.

**Do not relax `_Real`.** It is also the *support* of `Normal`, `Cauchy`, `StudentT`, `Laplace`, `Logistic` and (via `real_vector`) `MultivariateNormal`, so widening it would stop catching `Normal(loc=inf)` and infinite samples across a dozen distributions. That buys one `-inf` logit at a high price.

Instead add two parameter-domain constraints:

1. `constraints.extended_real`: `[-inf, inf]`, NaN rejected. For sigmoid-linked logits (`BernoulliLogits`, `BinomialLogits`) where the probs sibling already allows `p` in `{0, 1}`, and for `LeftTruncatedDistribution.low` / `RightTruncatedDistribution.high`.
2. `constraints.softmax_logits` (`event_dim = 1`): no NaN, no `+inf`, and at least one entry above `-inf`. In other words "the softmax is well defined". For `CategoricalLogits` and `MultinomialLogits`. This is exactly your intuition of admitting `-inf` but not `+inf`: `softmax([inf, 0.])` is `inf - inf` in the normalizer, and `softmax([inf, inf])` has a path-dependent limit, so there is no correct value to return. The all `-inf` vector is rejected for the same reason `simplex` rejects `[0, 0]`.

Then fix the `0 * inf = nan` sites that infinite logits now reach, otherwise loosening the constraint just converts a `ValueError` into a silent `nan`:

- `CategoricalLogits.entropy` returns `nan` today for `[-inf, 0.]` (the `probs` version was fixed in #2276, the logits one was missed)
- `BernoulliLogits.entropy`
- the `value * logits` term in `BinomialLogits.log_prob` and `MultinomialLogits.log_prob`, which makes `log_prob(0)` a `nan` instead of `0`

The pattern to follow is already in the repo, in `_to_logits_multinom`: sanitize the *operand* of the product with `jnp.where`, not the result, so gradients stay finite too.

Worth keeping out of this PR: `GeometricProbs` / `GeometricLogits` and `NegativeBinomialLogits` have their own pre-existing boundary problems (`GeometricProbs(1.0).log_prob(0)` returns `-inf` instead of `0.0`, and `NegativeBinomialLogits` with `+inf` implies a `GammaPoisson` rate of `0`, which is improper). Those deserve separate issues.

@jaberg , if you want some help or if you would rather want us to fix it please let us know :) Whatever work sfor you.

**Comment by jaberg** [comment]
Candidly, validate_args=False on my CategoricalLogits line is offering a pretty great solution for me right now relative to this suggested fix. It looks like a potential iceberg of changes... if y'all fix as much of this as you see fit, I'd appreciate it.

**Comment by jaberg** [comment]
Thanks for making numpyro by the way. I'd looked a jax a few times, but numpyro was what got me to make the leap. I'm kind of new to hierarchical Bayesian modelling, and I've found numpyro examples, docs, and source code to be a great way to learn about it.
